mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/ChannelPoolContract.java

```diff
@@ -0,0 +1,97 @@
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.events;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.time.Duration;
+import java.util.NoSuchElementException;
+
+import org.junit.jupiter.api.Test;
+
+import com.rabbitmq.client.Channel;
+
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.SignalType;
+import reactor.rabbitmq.ChannelPool;
+
+interface ChannelPoolContract {
+
+    ChannelPool getChannelPool(int poolSize);
+
+    @Test
+    default void channelPoolShouldCreateDifferentChannels() {
+        ChannelPool channelPool = getChannelPool(2);
+        Mono<? extends Channel> channelMono = channelPool.getChannelMono();
+        Channel channel1 = channelMono.block();
+        Channel channel2 = channelMono.block();
+
+        assertThat(channel1.getChannelNumber())
+            .isNotEqualTo(channel2.getChannelNumber());
+    }
+
+    @Test
+    default void channelPoolShouldReleaseChannelWhenRetunredToThePool() {
+        ChannelPool channelPool = getChannelPool(2);
+        Mono<? extends Channel> channelMono = channelPool.getChannelMono();
+        Channel channel1 = channelMono.block();
+        Channel channel2 = channelMono.block();
+
+        returnToThePool(channelPool, channel2);
+        Channel channelAfterReturned = channelMono.block();
+
+        assertThat(channelAfterReturned.getChannelNumber())
+            .isEqualTo(channel2.getChannelNumber());
+    }
+
+    @Test
+    default void channelPoolShouldWaitTillTheNextReleaseWhenAllChannelsAreTaken() {
+        ChannelPool channelPool = getChannelPool(2);
+        Mono<? extends Channel> channelMono = channelPool.getChannelMono();
+        Channel channel1 = channelMono.block();
+        Channel channel2 = channelMono.block();
+
+        Mono.delay(Duration.ofSeconds(2))
+            .doOnSuccess(any -> returnToThePool(channelPool, channel2))
+            .subscribe();
+
+        Channel channelAfterReturned = channelMono.block();
+        assertThat(channelAfterReturned.getChannelNumber())
+            .isEqualTo(channel2.getChannelNumber());
+    }
+
+    @Test
+    default void channelPoolShouldThrowAfterTimeoutWhenAllChannelsAreTaken() {
+        ChannelPool channelPool = getChannelPool(2);
+        Mono<? extends Channel> channelMono = channelPool.getChannelMono();
+        Channel channel1 = channelMono.block();
```

Comment:
I don't like that `block` is a actually a get. It's harder to read than necessary